llvm/lib/Target/AMDGPU/AMDGPUSplitModule.cpp

```diff
@@ -476,12 +486,36 @@ void SplitGraph::Node::visitAllDependencies(
   }
 }
 
+/// Checks if \p I has MD_callees and if it does, parse it and put the function
+/// in \p Callees.
+///
+/// \returns true if there was metadata and it was parsed correctly. false if
+/// there was no MD or if it contained unknown entries.
+static bool handleCalleesMD(const Instruction &I,
+                            SmallVector<Function *> &Callees) {
+  auto *MD = I.getMetadata(LLVMContext::MD_callees);
+  if (!MD)
+    return false;
+
+  for (const auto &Op : MD->operands()) {
+    Function *Callee = mdconst::extract_or_null<Function>(Op);
+    if (!Callee)
+      return false;
```

Comment:
I'm not sure about this return.

I find it strange that the code pushes some, but not all of the callees. Maybe I'm missing some context and the cases where we fail to extract the function are at the end.

Reply:
IIRC I copy pasted it from another file, but the idea is just that if the MD is ill-formed we conservatively give up and assume the worst (indirect call to anything).

When this function returns false, we just end up discarding KnownCallees entirely

I'll refactor this to make it a bit clearer

Reply:
It is not discarded. The for-loop continues iterating over the instructions and if it finds another indirect call with well-formed metadata it may push the callees before the ill-formed one in the set.

Reply:
> I find it strange that the code pushes some, but not all of the callees.

That is UB based on LLVM Lang.

Reply:
I think this might be more like an assertion.

Reply:
> I think this might be more like an assertion.

Both places that read this MD also check for null, see `ModuleSummaryAnalysis.cpp:476` and `AttributorAttributes.cpp:12215`

> It is not discarded. The for-loop continues iterating over the instructions and if it finds another indirect call with well-formed metadata it may push the callees before the ill-formed one in the set.

I don't mean the set is discarded entirely, but it's definitely never used afterwards. When `HasUnknownIndirectCall` is set we just don't use the set.
I'll try to make it clearer in code.

Reply:
I see now. I haven't noticed the interaction with `HasUnknownIndirectCall`